aten/src/ATen/native/mps/operations/Convolution.mm

```diff
@@ -485,7 +485,19 @@ Tensor mps_convolution_backward_weights(
           MPSGraphTensor* gradOutputTensor = native_mps::mpsGraphRankedPlaceHolder(mpsGraph, native_mps::getMPSScalarType(grad_output_t.scalar_type()), gradOutputShape);
           MPSGraphTensor* inputTensor = native_mps::mpsGraphRankedPlaceHolder(mpsGraph, input_t);
 
-          MPSGraphTensor *gradOutputTensorTranspose = gradOutputTensor;
+          auto input_mps_dtype = native_mps::getMPSDataType(input_t.scalar_type());
+          bool isFP16 = input_mps_dtype == MPSDataTypeFloat16;
+
+          MPSGraphTensor* gradOutputCastTensor = gradOutputTensor;
+          MPSGraphTensor* inputCastTensor = inputTensor;
+
+          if(isFP16) {
+            // up casting to fp32
+            gradOutputCastTensor = mps::castMPSTensor(mpsGraph, gradOutputTensor, MPSDataTypeFloat32);
```

Comment:
@Ronian526 - can we instead cast to FP16 here instead of FP32? Convolution is one of the ops where we don't really need the higher precision. I mean, could we run this whole graph through FP16 only (if input is FP16)?

Reply:
Denis this is Gradient with weights. This is the issue he is trying to work around.